Add Gradient descent algorithms

paquiteau opened the pull request:
As announced in #194 , this PR adds support for gradient descent algorithms with various gradient direction and speed preconditioning.

sfarrens commented:
Hi @paquiteau thanks for the suggestion and the PR! I will review this properly once all the CI issues are resolved. If you are having any problem with some of the [WPS](https://github.com/wemake-services/wemake-python-styleguide) errors let me know as we can relax some errors for certain files. In particular, it will be necessary for `algorithms/__init_.py` in order to support the `*` imports.

paquiteau commented:
I fixed most of the issues with linting, and updated  `setup.cfg` when I could not find another way.  I still have to write some test specific to the gradient descent though

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+        """Update the gradient descent direction."""
+        self._dir_grad = grad
+
+    def update_grad_speed(self, grad):
+        """Update the gradient descent speed."""
+        pass
```
I guess here you are simply defining a method you expect the child classes to override, right? I don't really think this is needed unless the child class is forced to override this (*e.g.* as an abstract method).

paquiteau replied:
I had to defined this empty method for the generic case. (cf `_update` method, which is commont to all gradient descent algorithms.

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+            metrics[obs.name] = obs.retrieve_metrics()
+        self.metrics = metrics
+
+    def reset(self):
+        """Reset internal state of the algorithm."""
+        pass
```
Same point regarding method overriding.

paquiteau replied:
same answer ;) some class reimplements it (AdaGrad for example), other not.

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+    """Generic Grad descent Algorithm with ADA acceleration scheme."""
+
+    def update_grad_speed(self, grad):
+        """Ada Acceleration Scheme."""
```
Is there a reference you can include for this?

sfarrens replied:
Should probably added a properties section as there is an input parameter (`grad`)

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+    -----
+    The gradient speed is defined as :
+
+    .. math:: s_k = \gamma s_{k-1}  + (1-\gamma) * |\nabla f|^2
```
Is there a reference you can include for this?

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+        self._gamma = gamma
+
+    def update_grad_speed(self, grad):
+        """Rmsprop update speed."""
```
Same point re: `grad`

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+    -----
+    The momentum update is defined as:
+
+    .. math:: m_k = \beta * m_{k-1} + \nabla f(x_k)
```
Is there a reference you can include for this?

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+        self._eps = 0
+
+    def update_grad_dir(self, grad):
+        """Momentum gradient direction update."""
```
Same point re: grad

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+    -----
+    The ADAM optimizer is defined as:
+
+    .. math::
+        m_{k+1} = \frac{1}{1-\beta^k}(\beta*m_{k}+(1-\beta)*|\nabla f_k|^2)
+    .. math::
+        s_{k+1} = \frac{1}{1-\gamma^k}(\gamma*s_k+(1-\gamma)*\nabla f_k)
```
Is there a reference you can include for this?

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+        self._gamma_pow = 1
+
+    def update_grad_dir(self, grad):
+        """ADAM Update of gradient direction."""
```
Same point re: grad

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+        )
+
+    def update_grad_speed(self, grad):
+        """ADAM Updatae of gradient speed."""
```
Same point re: grad

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+        )
+
+    def update_grad_dir(self, grad):
+        """SAGA Update gradient direction."""
```
Same point re: grad

sfarrens commented on modopt/tests/test_opt.py:
```diff
@@ -37,292 +37,6 @@ class Dummy(object):
     pass
 
 
```
excess blank line

sfarrens commented:
Hi @paquiteau!

I made a first pass on your PR. Firstly, very nice and meticulous work 👏 . 

- I proposed some minor changes to `gradient_descent.py` just to clean things up a bit, but otherwise the implementation looks good and the tests pass. The main thing you could add would be some references as it is good to have a direct comparison between a given equation in a given paper/book and the code implemented.
- It appears there are some minor conflicts with the `develop` branch. This should be easy enough to fix if you pull the latest commits from `develop`.

Once these points are resolved I will make a final pass, in particular for the moved algorithms, which I believe should not have changed much.

paquiteau commented:
Thanks for the review. 
I added the references where it was required, and documented what was not :) 

Also, I change the `update_grad_dir` and `_update_grad_speed` to private methods, these should not be changed/called by the user, as they are directly called inside `_update`.

sfarrens commented:
Hi @paquiteau great! I still see a couple or merge conflicts listed between your branch and `develop`.

```
Conflicting files
modopt/opt/algorithms.py
modopt/tests/test_opt.py
```

sfarrens commented on modopt/opt/algorithms/gradient_descent.py:
```diff
@@ -0,0 +1,354 @@
+
+        self.x_final = self._x_new
+
+    def _update(self):
```
@paquiteau this thread still needs to be resolved

sfarrens commented:
@paquiteau I have finished checking the other modules and, apart from some very minor things that I can clean up myself when I make the next release, everything looks good.

So, as soon as you resolve the last thread and we figure out the the conflict issue this PR can be merged! 👏
